llvm/lib/CodeGen/AsmPrinter/AsmPrinter.cpp

```diff
@@ -1401,17 +1406,48 @@ void AsmPrinter::emitBBAddrMapSection(const MachineFunction &MF) {
   uint8_t BBAddrMapVersion = OutStreamer->getContext().getBBAddrMapVersion();
   OutStreamer->emitInt8(BBAddrMapVersion);
   OutStreamer->AddComment("feature");
-  auto FeaturesBits = static_cast<uint8_t>(PgoAnalysisMapFeatures.getBits());
-  OutStreamer->emitInt8(FeaturesBits);
-  OutStreamer->AddComment("function address");
-  OutStreamer->emitSymbolValue(FunctionSymbol, getPointerSize());
-  OutStreamer->AddComment("number of basic blocks");
-  OutStreamer->emitULEB128IntValue(MF.size());
-  const MCSymbol *PrevMBBEndSymbol = FunctionSymbol;
+  auto Features = getBBAddrMapFeature(MF, MBBSectionRanges.size());
+  OutStreamer->emitInt8(Features.encode());
   // Emit BB Information for each basic block in the function.
+  if (Features.MultiBBRange) {
+    OutStreamer->AddComment("number of basic block ranges");
+    OutStreamer->emitULEB128IntValue(MBBSectionRanges.size());
+  }
+  // Number of blocks in each MBB section.
+  MapVector<unsigned, unsigned> MBBSectionNumBlocks;
+  const MCSymbol *PrevMBBEndSymbol = nullptr;
+  if (!Features.MultiBBRange) {
+    OutStreamer->AddComment("function address");
```

Comment:
Why we need to handle `MBBSectionRanges.size() == 1` specially? Can we have one format BBAddrMap?

Reply:
The encoding and decoding will be different. We are adding MultiBBRange as a feature. If the feature is disabled, we have a slightly smaller encoding. Although, as I have mentioned before it doesn't really matter how we encode these. The output format of BBAddrMap (ELFTypes.h) is oblivious to these features.

Reply:
How about doesn't add MultiBBRange as a feature? The default format is starting with `NumBBRanges` even if `NumBBRanges` is one. There are many codes which be added because of MultiBBRange feature.

Reply:
A second reason is backward and forward compatibility. As written, this PR does not change the encoding for the normal case. So we don't need to add a new version (which leads to even more technical debt).

Reply:
Thanks. I understand your point now.